src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1467,27 +1467,43 @@ internal void ClearTransaction()
 
         internal void PerformUserCancellation()
         {
-            _userCancellationRequested = true;
-
-            if (AttemptPostgresCancellation && SupportsPostgresCancellation)
+            lock (CancelLock)
             {
-                var cancellationTimeout = Settings.CancellationTimeout;
-                if (PerformPostgresCancellation() && cancellationTimeout >= 0)
+                if (IsBroken)
+                    return;
+
+                _userCancellationRequested = true;
+
+                if (AttemptPostgresCancellation && SupportsPostgresCancellation)
                 {
-                    if (cancellationTimeout > 0)
+                    var cancellationTimeout = Settings.CancellationTimeout;
+                    if (PerformPostgresCancellation() && cancellationTimeout >= 0)
                     {
-                        UserTimeout = cancellationTimeout;
-                        ReadBuffer.Timeout = TimeSpan.FromMilliseconds(cancellationTimeout);
-                        ReadBuffer.Cts.CancelAfter(cancellationTimeout);
+                        if (cancellationTimeout > 0)
+                        {
+                            lock (this)
+                            {
+                                if (IsBroken)
```

Comment:
@vonzshik 
Here is a snapshot from the the crash dump that shows the point of failure, the null ref and the State (closed):
![image](https://user-images.githubusercontent.com/22180333/105280024-dd2e4e00-5b76-11eb-9ee6-cb776cdc741d.png)

Reply:
Well, while this snapshot doesn't tell me much, if the connector is closed, then most likely we've started to request the cancellation while the query was executing, then it's cancelled/completed (while we're still in the cancellation logic). And after that, the connector is closed (might be due to a number of reasons, like pruning), and we get the nre. Updated PR to also fix this kind of behavior.

Reply:
Agreed, intercepted every "not connected" state now.